rpc/src/cluster_tpu_info.rs

```diff
@@ -64,6 +67,34 @@ impl TpuInfo for ClusterTpuInfo {
         }
         unique_leaders
     }
+
+    fn get_leader_tpus_with_slots(
+        &self,
+        max_count: u64,
+        protocol: Protocol,
+    ) -> Vec<(&SocketAddr, Slot)> {
+        let recorder = self.poh_recorder.read().unwrap();
+        let leaders: Vec<_> = (0..max_count)
+            .filter_map(|i| {
+                recorder.leader_and_slot_after_n_slots(i * NUM_CONSECUTIVE_LEADER_SLOTS)
+            })
+            .collect();
+        drop(recorder);
+        let mut unique_leaders = vec![];
+        let mut leader_slot: HashMap<&Pubkey, Slot> = HashMap::default();
+        for leader in leaders.iter() {
+            if let Some(addr) = self.recent_peers.get(&leader.0).map(|addr| match protocol {
+                Protocol::UDP => &addr.0,
+                Protocol::QUIC => &addr.1,
+            }) {
+                if !leader_slot.contains_key(&leader.0) {
+                    unique_leaders.push((addr, leader.1));
+                    leader_slot.insert(&leader.0, leader.1);
+                }
+            }
+        }
+        unique_leaders
+    }
```

Comment:
```suggestion
    fn get_leader_tpus_with_slots(
        &self,
        num_leaders: u64,
        protocol: Protocol,
    ) -> Vec<(&SocketAddr, Slot)> {
        let recorder = self.poh_recorder.read().unwrap();
        let leaders: Vec<_> = (0..num_leaders)
            .filter_map(|future_slot| {
                NUM_CONSECUTIVE_LEADER_SLOTS
                    .checked_mul(future_slot)
                    .and_then(|target_slot| {
                        recorder.leader_and_slot_after_n_slots(target_slot)
                    })
            })
            .collect();
        drop(recorder);
        let addrs_to_slots = leaders
            .into_iter()
            .filter_map(|(leader_id, leader_slot)| {
                self.recent_peers
                    .get(&leader_id)
                    .map(|(udp_tpu, quic_tpu)| match protocol {
                        Protocol::UDP => (udp_tpu, leader_slot),
                        Protocol::QUIC => (quic_tpu, leader_slot),
                    })
            })
            .collect<HashMap<_, _>>();
       // needs `use itertools::Itertools`
       addrs_to_slots.into_iter().sort_by_key(|(_addr, slot)| slot).collect<Vec<_>>()
    }
```

Reply:
I think the problem with the suggested change is when collecting to HashMap, it overwrites the existing keys with the smaller slot. I intended to get the smallest slot for a unique leader.

Reply:
that's fine, we can just build `leaders` in reverse
```
let future_slot = num_leaders.wrapping_sub(future_slot);
```
at the top of the closure

Reply:
done